### PR TITLE
Gracefully handling connection issues

### PR DESCRIPTION
**Pull request by ShanniLi**
r @anson627 

Several major updates:
1. gracefully handling outbound exceptions
2. gracefully handling inbound exceptions
3. reset connection on 10 continuous timeout
4. remove connection from peer when it fails
5. add benchmark for multiple connection

### CODE REVIEW

File: tchannel-core/src/test/java/com/uber/tchannel/channels/ConnectionTest.java [hunk 1]
```diff
@@ -135,15 +202,174 @@ public void testConnectionFailure() throws Exception {
...
+            this.delayed = delayed;
+        }
 
         @Override
         public RawResponse handle(Request request) {
+
+            if (delayed) {
+                try {
+                    sleep(300);
+                } catch (Exception ex) {
```
Comment (anson627) [inline tchannel-core/src/test/java/com/uber/tchannel/channels/ConnectionTest.java]: do nothing if exception?

File: tchannel-core/src/main/java/com/uber/tchannel/api/TChannel.java [hunk 2]
```diff
@@ -153,15 +153,24 @@ public SubChannel makeSubChannel(String service) {
         return this.makeSubChannel(service, Connection.Direction.NONE);
     }
 
-    public void shutdown(boolean sync) throws InterruptedException, ExecutionException {
+    public void shutdown(boolean sync) {
```
Comment (anson627) [inline tchannel-core/src/main/java/com/uber/tchannel/api/TChannel.java]: why logging instead of exception?

**Comment by anson627** [comment]
minor comments, otherwise lgtm

File: tchannel-core/src/test/java/com/uber/tchannel/channels/ConnectionTest.java (hunk 1 above)
Comment (ShanniLi) [inline tchannel-core/src/test/java/com/uber/tchannel/channels/ConnectionTest.java]: It is InterruptedException, not too much we can do. First, it is unlikely we get interrupted here. Second, even we get interrupted, the test will fail. So, no fasle positives.

File: tchannel-core/src/main/java/com/uber/tchannel/api/TChannel.java (hunk 2 above)
Comment (ShanniLi) [inline tchannel-core/src/main/java/com/uber/tchannel/api/TChannel.java]: Not sure it is a good idea to ask users to handle these two exceptions. They are shutting down tchannel, so I guess they won't care about these two exceptions at this stage.

**Pull request merged**